kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -0,0 +1,210 @@
+    private final Optional<SetTransaction> txnIdentifier;
+
+    private boolean committed; // To avoid trying to commit the same transaction again.
+
+    public TransactionImpl(
+            Path dataPath,
+            Path logPath,
+            SnapshotImpl readSnapshot,
+            String engineInfo,
+            String operation,
+            Optional<StructType> newSchema,
+            Protocol protocol,
+            Metadata metadata,
+            Optional<Long> readVersion,
+            Optional<Predicate> readPredicate,
+            Optional<SetTransaction> txnIdentifier) {
+        this.dataPath = dataPath;
+        this.logPath = logPath;
+        this.readSnapshot = readSnapshot;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+        this.newSchema = newSchema;
+        this.protocol = protocol;
+        this.metadata = metadata;
+        this.readVersion = readVersion;
+        this.readPredicate = readPredicate;
+        this.txnIdentifier = txnIdentifier;
+    }
+
+    @Override
+    public Row getState(TableClient tableClient) {
+        StructType tableSchema = readSnapshot.getSchema(tableClient);
+        // Physical equivalent of the logical read schema.
+        StructType physicalReadSchema = ColumnMapping.convertToPhysicalSchema(
+                tableSchema,
+                tableSchema,
+                ColumnMapping.getColumnMappingMode(metadata.getConfiguration()));
+
+        return TransactionStateRow.of(
+                metadata,
+                protocol,
+                tableSchema.toJson(),
+                physicalReadSchema.toJson(),
+                dataPath.toString());
+    }
+
+    @Override
+    public List<String> getPartitionColumns(TableClient tableClient) {
+        throw new UnsupportedOperationException("Not yet implemented");
+    }
+
+    @Override
+    public StructType getSchema(TableClient tableClient) {
+        return newSchema.orElseGet(() -> readSnapshot.getSchema(tableClient));
+    }
+
+    @Override
+    public TransactionCommitStatus commitWithRetries(
+            TableClient tableClient,
+            CloseableIterable<Row> stagedData,
+            int maxRetries)
+            throws ConcurrentWriteException {
+        checkState(!committed, "Transaction is already committed. Create a new transaction.");
+        List<Row> metadataActions = new ArrayList<>();
+        metadataActions.add(createCommitInfoSingleAction(generateCommitAction()));
+        metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+        metadataActions.add(createProtocolSingleAction(protocol.toRow()));
+        addSetTransactionIfPresent(metadataActions);
+
+        CloseableIterator<Row> stageDataIter = stagedData.iterator();
+
+        // Create a new CloseableIterator that will return the commit action, metadata action, and
+        // the stagedData
+        CloseableIterator<Row> stagedDataWithCommitInfo = new CloseableIterator<Row>() {
```

Comment:
are you basically just unioning two iterators? is there a simpler util method we can use to do that instead?